src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1206,58 +1206,79 @@ public override BoundNode VisitNullCoalescingOperator(BoundNullCoalescingOperato
         {
             Debug.Assert(!IsConditionalState);
 
-            VisitRvalue(node.LeftOperand);
-            if (IsConstantNull(node.LeftOperand))
+            BoundExpression leftOperand = node.LeftOperand;
+            BoundExpression rightOperand = node.RightOperand;
+
+            Result leftResult = VisitRvalueWithResult(leftOperand);
+            if (IsConstantNull(leftOperand))
             {
-                VisitRvalue(node.RightOperand);
+                VisitRvalue(rightOperand);
                 return null;
             }
 
             var leftState = this.State.Clone();
-            var leftType = _result.Type;
-            if (leftType?.IsNullable == false)
+            if (leftResult.Type?.IsNullable == false)
             {
-                ReportStaticNullCheckingDiagnostics(ErrorCode.HDN_ExpressionIsProbablyNeverNull, node.LeftOperand.Syntax);
+                ReportStaticNullCheckingDiagnostics(ErrorCode.HDN_ExpressionIsProbablyNeverNull, leftOperand.Syntax);
             }
 
-            bool leftIsConstant = node.LeftOperand.ConstantValue != null;
+            bool leftIsConstant = leftOperand.ConstantValue != null;
             if (leftIsConstant)
             {
                 SetUnreachable();
             }
 
-            leftType = InferResultNullability(node.LeftOperand, node.LeftConversion, node.Type, leftType);
-
-            VisitRvalue(node.RightOperand);
-            var rightType = _result.Type;
-
+            var rightResult = VisitRvalueWithResult(rightOperand);
             IntersectWith(ref this.State, ref leftState);
 
-            TypeSymbolWithAnnotations resultType;
-
-            if (node.Type.IsErrorType())
+            bool? getIsNullable(BoundExpression e, Result r) => (r.Type is null) ? e.IsNullable() : r.Type.IsNullable;
+            bool? resultIsNullable = getIsNullable(leftOperand, leftResult) == false ? false : getIsNullable(rightOperand, rightResult);
+            TypeSymbol getLeftResultType(TypeSymbol leftType, TypeSymbol rightType)
```

Comment:
Would it be ok to move the local functions at the end of the method? Helps maintaining the flow of main logic. Thanks